src/librustc_trans/back/linker.rs

```diff
@@ -69,48 +74,107 @@ impl<'a> GnuLinker<'a> {
     }
 }
 
+impl<'a> fmt::Debug for GnuLinker<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.cmd.fmt(f)
+    }
+}
+
+impl<'a> fmt::Display for GnuLinker<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.name.fmt(f)
+    }
+}
+
 impl<'a> Linker for GnuLinker<'a> {
-    fn link_dylib(&mut self, lib: &str) { self.cmd.arg("-l").arg(lib); }
-    fn link_staticlib(&mut self, lib: &str) { self.cmd.arg("-l").arg(lib); }
-    fn link_rlib(&mut self, lib: &Path) { self.cmd.arg(lib); }
-    fn include_path(&mut self, path: &Path) { self.cmd.arg("-L").arg(path); }
-    fn framework_path(&mut self, path: &Path) { self.cmd.arg("-F").arg(path); }
-    fn output_filename(&mut self, path: &Path) { self.cmd.arg("-o").arg(path); }
-    fn add_object(&mut self, path: &Path) { self.cmd.arg(path); }
-    fn position_independent_executable(&mut self) { self.cmd.arg("-pie"); }
-    fn args(&mut self, args: &[String]) { self.cmd.args(args); }
+    fn link_dylib(&mut self, lib: &str) {
+        if self.is_lld && self.sess.target.target.options.is_like_osx {
+            let mut v = OsString::from("-l");
+            v.push(lib);
+            self.arg(&v);
+        } else {
+            self.arg("-l").arg(lib);
+        }
+    }
+    fn link_staticlib(&mut self, lib: &str) {
+        if self.is_lld && self.sess.target.target.options.is_like_osx {
+            let mut v = OsString::from("-l");
+            v.push(lib);
+            self.arg(&v);
+        } else {
+            self.arg("-l").arg(lib);
+        }
+    }
+    fn link_rlib(&mut self, lib: &Path) { self.arg(lib); }
+    fn include_path(&mut self, path: &Path) { self.arg("-L").arg(path); }
+    fn framework_path(&mut self, path: &Path) { self.arg("-F").arg(path); }
+    fn output_filename(&mut self, path: &Path) { self.arg("-o").arg(path); }
+    fn add_object(&mut self, path: &Path) { self.arg(path); }
+    fn position_independent_executable(&mut self) { self.arg("-pie"); }
+    fn arg<S: AsRef<OsStr>>(&mut self, arg: S) -> &mut Self {
+        if self.is_lld {
+            match arg.as_ref().to_str() {
+                Some(args_str) => {
+                    for arg_str in args_str.trim_left_matches("-Wl,").split(',') {
```

Comment:
This is the part I don't follow -- it doesn't seem reasonable to drop flags like `-m64` on the floor, so the question is really: what is gcc doing with this flag, and what is the linker flag that it translates to?

We should be able to apply this treatment to (almost) all these flags and then separate into two arrays as you suggest (since there are a few flags used specifically to disable some of gcc's behaviour, e.g. `-nodefaultlibs`.